src/common/io/io.cpp

```diff
@@ -573,6 +584,53 @@ error::Error FileReader::Rewind() {
 	return error::NoError;
 }
 
+ExpectedSize BufferedReader::Read(vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) {
+	Vsize bytes_read_buffer {0};
+	Vsize bytes_read_wrapper {0};
+
+	if (!rewind_done_) {
```

Comment:
I agree with the general approach here, but to me the logic seems backwards, and I have some suggestions:

1. Instead of checking for `!rewind_done_`, check for `rewind_done_`, and then do the buffered reading in there. Don't bother concatenating the two sources, this is IMHO not necessary. All consumers support short reads.
2. As soon as it returns `0`, set `rewind_done_ = false` and fall through . From there you follow up with the `wrapped_reader`.
3. (Bonus): If you also implement `ByteReader::Rewind() { bytes_read_ = 0; }`, then you can call it from `BufferedReader::Rewind()` which will allow `Rewind()` to be used multiple times, as long as `StopBuffering()` hasn't been called. I know, this isn't necessary, but I would argue it improves code quality as well.
4. And finally, if `!rewind_done_ && stop_done_`, then you should call `buffer_.clear()` to save memory.